File: src-2/producer-consumer.go
package main

import (
	"fmt"
	"runtime"
	"sync"
)

const (
	producers = 1000
	consumers = 1000
	capacity  = 10
)

func main() {
	// Set the number of CPU cores used by the program to 1
	runtime.GOMAXPROCS(1)

	var wg sync.WaitGroup
	wg.Add(producers + consumers)

	buffer := make(chan int, capacity)

	// Launch producer Goroutines
	for i := 0; i < producers; i++ {
		go func(id int) {
			defer wg.Done()
			buffer <- id
			fmt.Printf("Producer %d produced\n", id)
		}(i)
	}

	// Launch consumer Goroutines
	for i := 0; i < consumers; i++ {
		go func(id int) {
			defer wg.Done()
			value := <-buffer
			fmt.Printf("Consumer %d consumed value %d\n", id, value)
		}(i)
	}

	wg.Wait()
	fmt.Println("Done")
}

File: src-2/producer-consumer.py
import os
import asyncio
from asyncio import Queue

os.environ["PYTHONASYNCIODEBUG"] = "1"

producers = 1000
consumers = 1000
capacity = 10

async def producer(queue, id):
    await queue.put(id)
    print(f"Producer {id} produced")

async def consumer(queue, id):
    value = await queue.get()
    print(f"Consumer {id} consumed value {value}")

async def main():
    queue = Queue(capacity)
    tasks = []

    # Launch producer tasks
    for i in range(producers):
        tasks.append(asyncio.create_task(producer(queue, i)))

    # Launch consumer tasks
    for i in range(consumers):
        tasks.append(asyncio.create_task(consumer(queue, i)))

    await asyncio.gather(*tasks)

asyncio.run(main())
